haystack/core/pipeline/utils.py

```diff
@@ -8,45 +8,40 @@
 from typing import Any, List, Optional, Tuple
 
 from haystack import logging
+from haystack.core.component import Component
+from haystack.tools import Tool, Toolset
 
 logger = logging.getLogger(__name__)
 
 
-def deepcopy_with_fallback(obj: Any, max_depth: Optional[int] = 100) -> Any:
+def deepcopy_with_fallback(obj: Any) -> Any:
```

Comment:
Nitpicks:
1. Do you think that `deepcopy_with_fallback` is still an explanatory name? Or can you think of better options?
2. Types: `Any` does not give any guarantees in terms of static type checking.
    What about something like the following?
    ```python
         from typing import TypeVar, List

         T = TypeVar('T')
         def deepcopy_with_fallback(obj: T) -> T:
    ``` 
    [Relevant StackOverflow thread](https://stackoverflow.com/questions/39817081/typing-any-vs-object)

Reply:
I agree on the types!

As for name I updated to `_deepcopy_with_exceptions`. Is that more clear? I also made it private so we can easily rename/change in the future